File: DSA/Array/0063_Unique_Paths_II/solution.php
// 0063. Unique Paths II
// Difficulty : Medium
// Tags       : array, dynamic-programming, matrix
// Solved on  : 2026-05-03
// LeetCode   : https://leetcode.com/problems/unique-paths-ii/

class Solution {

    private $dp = [];

    function __construct(){
        for($i = 0; $i <=101;$i++){
            $this->dp[$i] = array_fill(0,101,-1);
        }
    }
    /**
     * @param Integer[][] $obstacleGrid
     * @return Integer
     */
    function uniquePathsWithObstacles($obstacleGrid) {
        $m = count($obstacleGrid);
        $n = count($obstacleGrid[0]);
        return $this->dfs(0,0,$obstacleGrid,$m,$n);
    }

    private function dfs($i,$j,$grid,$m,$n){
        if ($i >= $m || $j >= $n) return 0;
        if ($grid[$i][$j] == 1) return 0;
        if ($i == $m - 1 && $j == $n -1 ) return 1;
        if ($this->dp[$i][$j] != -1 ) return $this->dp[$i][$j];
        return $this->dp[$i][$j] = $this->dfs($i+1,$j,$grid,$m,$n) + $this->dfs($i,$j+1,$grid,$m,$n);
    }
}

File: DSA/Array/0063_Unique_Paths_II/solution.cpp
// 0063. Unique Paths II
// Difficulty : Medium
// Tags       : array, dynamic-programming, matrix
// Solved on  : 2019-11-28
// LeetCode   : https://leetcode.com/problems/unique-paths-ii/

class Solution {
public:
    int dp[110][110];
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        memset(dp,-1,sizeof(dp));
        return f(0,0,obstacleGrid);
    }
    
    int f(int i,int j,vector<vector<int>>& obstacleGrid){
        int m=obstacleGrid.size();
        int n=0;
        if(m>0){
            n=obstacleGrid[0].size();
        }
        
        if(i>=m or j>=n){
            return 0;
         }
        if(obstacleGrid[i][j]==1){
            return 0;
        }
        if(i==m-1 and j==n-1 && obstacleGrid[i][j]!=1){
            return 1;
        }
        if(dp[i][j]!=-1){
            return dp[i][j];
        }
        int a=0;
        if(i+1 <m && j<n && obstacleGrid[i+1][j]!=1){
            a+=f(i+1,j,obstacleGrid);
        }
        if(i <m && j+1<n && obstacleGrid[i][j+1]!=1){
            a+=f(i,j+1,obstacleGrid);
        }
        
        return dp[i][j]=a;
    }
};